File: backend/app/routers/query_validation.py
"""
[AI GENERATED]
Model: GitHub Copilot (GPT-4.1)
Logic: Implements Advanced Query Validation & BI Integration API. Provides endpoints to validate SQL/BI queries, return errors/warnings, and simulate BI integration checks. Supports query linting, permission checks, and connector compatibility.
Why: Enables safe, compliant, and compatible query execution for analytics and BI tools.
Root Cause: No API for advanced query validation or BI integration checks.
Context: Used by backend for query validation, and by frontend for admin/analyst query UI. Future: extend for live query execution, schema introspection, and BI tool-specific validation.
Model Suitability: GPT-4.1 is suitable for FastAPI validation APIs; for advanced query analysis, consider Claude 3 or Gemini 1.5.
"""

from fastapi import APIRouter, Depends, HTTPException, Body
from sqlmodel import Session
from ..dependencies import get_session, require_admin
from typing import List, Optional

router = APIRouter(prefix="/v1/query_validation", tags=["Advanced Query Validation & BI Integration"])

# --- API Endpoints ---
@router.post("/validate", dependencies=[Depends(require_admin)])
async def validate_query(
    query: str = Body(...),
    connector: Optional[str] = Body(None),
):
    # Simulate validation (basic checks for demo)
    errors = []
    warnings = []
    if not query.strip().lower().startswith("select"):
        errors.append("Only SELECT queries are allowed.")
    if "drop" in query.lower() or "delete" in query.lower():
        errors.append("Destructive queries are not permitted.")
    if connector and connector not in ["powerbi", "tableau", "looker"]:
        warnings.append(f"Connector '{connector}' is not officially supported.")
    return {"errors": errors, "warnings": warnings, "valid": not errors}

@router.post("/bi_check", dependencies=[Depends(require_admin)])
async def bi_integration_check(
    query: str = Body(...),
    connector: str = Body(...),
):
    # Simulate BI integration check (always success for demo)
    return {"compatible": True, "connector": connector}

File: src/frontend/src/services/queryValidation.js
// [AI GENERATED]
// Model: GitHub Copilot (GPT-4.1)
// Logic: Service for Advanced Query Validation & BI Integration API. Methods to validate queries and check BI integration compatibility.
// Why: Enables frontend admin/analyst UI to validate queries and check BI tool compatibility.
// Root Cause: No frontend integration for new backend query validation API.
// Context: Used by admin/analyst UI for query validation and BI integration. Future: extend for live query execution, schema introspection, and BI tool-specific validation.
// Model Suitability: GPT-4.1 is suitable for REST API service patterns; for advanced query analysis, consider Claude 3 or Gemini 1.5.

import axios from 'axios';

const BASE_URL = '/v1/query_validation';

const QueryValidationService = {
    validateQuery: (payload) => axios.post(`${BASE_URL}/validate`, payload),
    biIntegrationCheck: (payload) => axios.post(`${BASE_URL}/bi_check`, payload),
};

export default QueryValidationService;
